From c1484511da3d5b69d5526af8fa5bf1233a4a0939 Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Fri, 11 Nov 2011 16:38:58 +0200
Subject: [PATCH] Fixed opengl (ftgl) font issue

Removed unnecessary state checks and modifications.Resolves an
issue where flickering could be seen on the rendered text when
there was no shaded surface present on the scene (automatic
z buffer management related).
---
 src/OpenGl/OpenGl_FontMgr.cxx |   40 +++++++++++++---------------------------
 1 files changed, 13 insertions(+), 27 deletions(-)

Index: oce/src/OpenGl/OpenGl_FontMgr.cxx
===================================================================
--- oce.orig/src/OpenGl/OpenGl_FontMgr.cxx
+++ oce/src/OpenGl/OpenGl_FontMgr.cxx
@@ -329,40 +329,26 @@ void OpenGl_FontMgr::render_text( const
     glPushMatrix();
 
     glScalef( _XCurrentScale, _YCurrentScale, 1 );
-    glPushAttrib( GL_ENABLE_BIT );
+    static GLint param;// = new GLint;
 
-    GLboolean enableTexture = glIsEnabled(GL_TEXTURE_2D);
-    GLboolean enableDepthTest = glIsEnabled(GL_DEPTH_TEST);
+    glGetTexEnviv(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, &param);
+    glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE);
 
-    if( !enableTexture )
-      glEnable(GL_TEXTURE_2D);
-    if ( !is2d ) {
-      if ( !enableDepthTest )
-        glEnable(GL_DEPTH_TEST);
-    } 
-    else if ( enableDepthTest ) {
-        glDisable(GL_DEPTH_TEST);
-    }
+    GLboolean enableAlpha = glIsEnabled(GL_ALPHA_TEST);
 
-    GLint* param = new GLint;    
-    glGetTexEnviv(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, param);
+    if (!enableAlpha) {
+      glAlphaFunc(GL_GEQUAL, 0.285f);
+      glEnable(GL_ALPHA_TEST);
+    }
 
-    glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE); 
-    glAlphaFunc(GL_GEQUAL, 0.285f);    
-    glEnable(GL_ALPHA_TEST);   
     OGLFont_Cache cache = _FontCache.Find( id );
     cache.Font->Render( text );
 
-    glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, *param);
-
-    if( !enableTexture )
-      glDisable(GL_TEXTURE_2D);
-    if( !enableDepthTest )
-      glDisable(GL_DEPTH_TEST);
+    glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, param);
 
-    delete param;
+    if( !enableAlpha )
+      glDisable(GL_ALPHA_TEST);
 
-    glPopAttrib();
     glMatrixMode( GL_MODELVIEW );
     glPopMatrix();
   }
